content/learn/beginner/inputs-1.md

```diff
@@ -0,0 +1,169 @@
+chapter, we'll learn how to create custom commands that take inputs.
+
+## Parsing Arguments
+
+Because parsing arguments is no easy task, we provide a convenience function, `parseArgs`, to parse the arguments passed
+to a custom command. For simple commands, this is the recommended way. In a later chapter, we will explore a more
+hands-on approach to parsing arguments.
+
+### Defining Arguments
+
+The first step is to define the arguments that the command will take. This is done using the aforementioned `parseArgs`
+function. The syntax is as follows:
+
+```go
+{{ $args := parseArgs <"num required args"> <"error message"> ...carg }}
+```
+
+The first argument `num required args` is the number of required arguments. You can use this to tell `parseArgs` how
+many arguments of the given definitions at the end are required. If the number of defined arguments is greater than the
+number of required arguments, the remaining arguments are considered optional. You cannot define optional positional
+arguments using `parseArgs`.
+
+After that, we can provide a custom error message that will be displayed if the arguments are not valid. Passing an
+empty string `""` will use a generated one based on the defined arguments. This is useful for providing more context to
+the user about what went wrong.
+
+The `...carg` is a variadic argument, that is, it can take any number of arguments.
+Each `carg` is a single argument definition. The `carg` function has the following syntax:
+
+```go
+{{ carg <"type"> <"description"> }}
+```
+
+Following types are supported:
+
+- `int` (whole number)
+- `float` (decimal number)
+- `string` (text)
+- `user` (user mentions, resolves to the [user](https://docs.yagpdb.xyz/reference/templates#user) structure)
+- `userid` (mentions or user IDs, resolves to the ID itself)
+- `member` (mentions or user IDs, resolves to the [member](https://docs.yagpdb.xyz/reference/templates#member)
+ structure)
+- `channel` (channel mention or ID, resolves to the channel structure)
+- `role` (role name or ID, resolves as type _\*discordgo.Role_)
+- `duration` (duration that is human-readable, i.e `10h5m` or `10 hour 5 minutes` would both resolve to the same
+   duration)
+
+The `description` is a human-readable description of the argument. This is used in the error message if the argument is
+not valid.
+
+Combining all of this, let's create a custom command that takes two arguments: a coolness level and a user to apply it
+to.
+
+```go
+{{ $args := parseArgs 2 "" (carg "int" "coolness level") (carg "member" "target member") }}
+```
+
+Note that we are using the `member` type here --- in general, it is recommended to use the `member` type for
+user-related arguments, as it validates that the target user is also a member of the server. There are very limited
+use-cases for `user` and `userid`, which will be left as an exercise to the reader.
```

Comment:
I do not really agree with the "very limited use-cases" part. If one just needs the user object, use a `user` carg. (It will validate that the user is a member of the server as well.) If one needs the member object, on the other hand, use the `member` carg. Finally, if one only needs a general user ID (without regard for whether it is valid), `userid` has the same effect as `int` but is self-documenting.

Reply:
Will rewrite this section to be more friendly towards `user` and
`userid` type arguments :-)